### PR TITLE
Add gdb convenience functions for working with heap and other randomised section addresses

### PR DESCRIPTION
**Pull request by wbowling**
### Description/Motivation/Screenshots ###
Quite often when debugging we need to print out the heap or an offset from the heap, and having to run `vmmap` to get the base address and copy/paste can be annoying. [GDB convenience functions](https://sourceware.org/gdb/onlinedocs/gdb/Functions-In-Python.html#Functions-In-Python) are just like convenience variables and can be used as arguments to other commands:

image

This PR adds functions for (offset is optional):
* `$_heap(offset)`
* `$_pie(offset)`
* `$_stack(offset)`
* `$_bss(offset)`
* `$_got(offset)`

 Happy to change or discuss the naming convention for these, this is just how the gdb built-in functions are (eg `$_strlen(str)`)

There's no doco or tests yet, just wanted to get any initial feedback first and then I'll make any changes and add them.

### How Has This Been Tested? ###

| Architecture | Yes/No                   | Comments               |
|--------------|:------------------------:|------------------------|
| x86-32       | :heavy_multiplication_x: |  |
| x86-64       | :heavy_check_mark: |                        |
| ARM          | :heavy_multiplication_x: |                        |
| AARCH64      | :heavy_multiplication_x: |                        |
| MIPS         | :heavy_multiplication_x: |                        |
| POWERPC      | :heavy_multiplication_x: |                        |
| SPARC        | :heavy_multiplication_x: |                        |
| `make tests` | :heavy_multiplication_x: |                        |

### Checklist ###

- [x] My code follows the code style of this project.
- [x] My change includes a change to the documentation, if required.
- [x] My change adds tests as appropriate.
- [x] I have read and agree to the **CONTRIBUTING** document.

---
This change is [Reviewable](https://reviewable.io/reviews/hugsy/gef/339)

### CODE REVIEW

File: gef.py [hunk 1]
```diff
@@ -2530,6 +2531,13 @@ def process_lookup_path(name, perm=Permission.ALL):
 
     return None
 
+def file_lookup_name_path(name, path):
+    """Look up for a file by its name and path.
```
Comment (Grazfather) [inline gef.py]: Weird wording

File: gef.py [hunk 2]
```diff
@@ -3582,6 +3590,11 @@ def register_priority_command(cls):
     __commands__.insert(0, cls)
     return cls
 
+def register_function(cls):
+    """Decorator for registering new GEF (sub-)command to GDB."""
```
Comment (Grazfather) [inline gef.py]: function, not command.

File: gef.py [hunk 3]
```diff
@@ -8603,6 +8616,81 @@ def get_settings_path(self):
...
+                return section.page_start
+        elif self._zone_:
+            zone = file_lookup_name_path(self._zone_, get_filepath())
+            if zone:
+                return zone.zone_start
+        return None
+
+@register_function
+class StackOffsetFunction(GenericOffsetFunction):
+    """Returns the current stack base address plus the given offset"""
```
Comment (Grazfather) [inline gef.py]: imperative tone and end with period (Return~s~)

File: gef.py [hunk 4]
```diff
@@ -8603,6 +8616,81 @@ def get_settings_path(self):
...
+
```
Comment (Grazfather) [inline gef.py]: nit but let's nix these extra lines for these functions

File: gef.py [hunk 5]
```diff
@@ -8730,6 +8818,10 @@ def is_loaded(x):
         # sort by command name
         self.loaded_commands = sorted(self.loaded_commands, key=lambda x: x[1]._cmdline_)
 
+        # load all of the functions
```
Comment (Grazfather) [inline gef.py]: Add "... and functions" to the docstring of `load`

File: gef.py (hunk 1 above)
Comment (wbowling) [reply gef.py]: copypasta

**Comment by wbowling** [comment]
The last commit adds these functions to the gef help command:

image

They look a bit strange though, what do you think? Will add something to the regular docs too but thought it might need to go here too

**Comment by wbowling** [comment]
> They look a bit strange though, what do you think? 

Could maybe add another separate command that just lists all the convenience functions? So that command will show up in help and if you run it it will print out more details about each

**Comment by wbowling** [comment]
Ok added
* a new command to list the functions
* test for new command and for the functions
* doco

File: gef.py [hunk 6]
```diff
@@ -8603,6 +8616,118 @@ def get_settings_path(self):
...
+        syntax = getattr(function, "_syntax_", "").lstrip()
+        w = max(1, get_terminal_size()[1] - 29)  # use max() to avoid zero or negative numbers
+        msg = "{syntax:<25s} -- {help:{w}s}".format(syntax=syntax, help=Color.greenify(doc), w=w)
+        self.docs.append(msg)
+        return
+
+    def do_invoke(self, argv):
+        self.dont_repeat()
+        gef_print(titlify("GEF - Convenience Functions"))
+        gef_print("These functions can be used as arguments to other commands to dynamically calculate values, eg: {:s}\n"
```
Comment (Grazfather) [inline gef.py]: This is annoying but we keep lines under 120 chars (easy to tell since that's max width github shows)

File: tests/runtests.py [hunk 7]
```diff
@@ -501,6 +508,61 @@ def test_func_get_pid(self):
...
+        self.assertFailIfInactiveSession(gdb_run_cmd(cmd))
+        res = gdb_start_silent_cmd(cmd)
+        self.assertNoException(res)
+        self.assertIn(b"\\177ELF", res)
+
+        cmd = "x/s $_pie(1)"
+        res = gdb_start_silent_cmd(cmd)
+        self.assertNoException(res)
+        self.assertNotIn(b"\\177ELF", res)
+        self.assertIn(b"ELF", res)
```
Comment (Grazfather) [inline tests/runtests.py]: clever

**Comment by Grazfather** [comment]
See https://circleci.com/gh/hugsy/gef/56

**Comment by wbowling** [comment]
> @wbowling looks like you forgot to add the bss.c

Ah whoops, sorry about that. Added and pushed :)

**Pull request merged**